include/rmm/mr/device/per_device_resource.hpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "default_memory_resource.hpp"
+#include "device_memory_resource.hpp"
+
+#include <mutex>
+#include <unordered_map>
+
+namespace rmm {
+namespace mr {
+
+/**
+ * @brief Strong type for a CUDA device identifier.
+ *
+ */
+struct cuda_device_id {
+  using value_type = int;
+
+  /**
+   * @brief Construct a `cuda_device_id` from the specified integer value
+   *
+   */
+  explicit constexpr cuda_device_id(value_type id) noexcept : id_{id} {}
+
+  /**
+   * @brief Returns the wrapped value
+   *
+   */
+  constexpr value_type value() const noexcept { return id_; }
+
+ private:
+  value_type id_;
+};
+
+namespace detail {
+inline std::mutex& map_lock()
+{
+  static std::mutex map_lock;
+  return map_lock;
+}
+
+inline auto& get_map()
+{
+  static std::unordered_map<cuda_device_id::value_type, device_memory_resource*>
```

Comment:
Since the number of devices is likely to be small, `std::map` is often faster in my experience. 
https://quick-bench.com/q/70Eg-tVt3M4lR4tlMqv3_ZiOcYs

(It's significantly faster in that test up through 64 items, and still barely faster at 128 items. Slower past 256.)